### PR TITLE
fix exprt::opX accesses in java_bytecode

### PR DESCRIPTION
**Pull request by kroening**
This improves type safety.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- n/a My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by NathanJPhillips — APPROVED** [review]

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 1]
```diff
@@ -916,8 +916,9 @@ void add_java_array_types(symbol_tablet &symbol_table)
 
     side_effect_exprt inc(ID_assign, typet(), location);
     inc.operands().resize(2);
-    inc.op0()=index_symexpr;
-    inc.op1()=plus_exprt(index_symexpr, from_integer(1, index_symexpr.type()));
+    to_binary_expr(inc).op0() = index_symexpr;
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp]: Let's introduce `side_effect_expr_assignt` instead?

File: jbmc/src/java_bytecode/java_bytecode_instrument.cpp [hunk 2]
```diff
@@ -494,20 +493,23 @@ optionalt<codet> java_bytecode_instrumentt::instrument_expr(const exprt &expr)
     {
       // this corresponds to a throw and so we check that
       // we don't throw null
-      result.add(check_null_dereference(expr.op0(), expr.source_location()));
+      result.add(check_null_dereference(
+        to_unary_expr(expr).op(), expr.source_location()));
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_instrument.cpp:497]: We have `side_effect_expr_throwt`

File: jbmc/src/java_bytecode/java_bytecode_instrument.cpp [hunk 3]
```diff
@@ -494,20 +493,23 @@ optionalt<codet> java_bytecode_instrumentt::instrument_expr(const exprt &expr)
...
     }
     else if(statement==ID_java_new_array)
     {
       // this corresponds to new array so we check that
       // length is >=0
-      result.add(check_array_length(expr.op0(), expr.source_location()));
+      result.add(check_array_length(
+        to_multi_ary_expr(expr).op0(), expr.source_location()));
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_instrument.cpp:504]: Add a container for java_new_array?

**Review by smowton — COMMENTED** [review]
Mostly good, I don't see that we're improving anything by using an unchecked cast to `multi_ary_exprt` and then using the similarly unguarded `op0` etc though, I think we should introduce appropriate containers for these

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d85e24f7b551f718d6b95143a9dbfc06fc562b29).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123241525

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp (hunk 1 above)
Comment (kroening) [reply jbmc/src/java_bytecode/java_bytecode_convert_class.cpp]: done

File: jbmc/src/java_bytecode/java_bytecode_instrument.cpp (hunk 2 above)
Comment (kroening) [reply jbmc/src/java_bytecode/java_bytecode_instrument.cpp:497]: yes, but that doesn't have operands -- will make that a separate PR.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 185a3c3b67ab84222f22e3006226f3ed6852de83).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123355159

**Pull request merged**